File: src/ros2_bridge/mappers/foxglove_msgs/frame_transforms.rs
//! Typed mapper for `foxglove_msgs/msg/FrameTransforms`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn frame_transforms_to_bus(msg: ros_env::foxglove_msgs::msg::FrameTransforms) -> crate::foxglove_msgs::msg::v1::FrameTransforms {
    crate::foxglove_msgs::msg::v1::FrameTransforms {
        transforms: msg.transforms.into_iter().map(crate::ros2_bridge::mappers::foxglove_msgs::frame_transform::frame_transform_to_bus).collect(),
    }
}

pub(crate) fn frame_transforms_to_ros(bus: crate::foxglove_msgs::msg::v1::FrameTransforms) -> ros_env::foxglove_msgs::msg::FrameTransforms {
    ros_env::foxglove_msgs::msg::FrameTransforms {
        transforms: bus.transforms.into_iter().map(crate::ros2_bridge::mappers::foxglove_msgs::frame_transform::frame_transform_to_ros).collect(),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct FoxgloveMsgsFrameTransformsMapper;

impl TypedTopicMapper for FoxgloveMsgsFrameTransformsMapper {
    type Ros = ros_env::foxglove_msgs::msg::FrameTransforms;
    type Bus = crate::foxglove_msgs::msg::v1::FrameTransforms;

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(frame_transforms_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(frame_transforms_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/foxglove_msgs/frame_transforms.py
"""Generated mapper for `foxglove_msgs/msg/FrameTransforms`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.foxglove_msgs.frame_transform import frame_transform_to_bus, frame_transform_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.foxglove_msgs.msg.v1 import FrameTransforms as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def frame_transforms_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.transforms.extend([frame_transform_to_bus(x) for x in msg.transforms])
    return bus


def frame_transforms_to_ros(bus):
    from foxglove_msgs.msg import FrameTransforms as RosMsg

    out = RosMsg()
    out.transforms = [frame_transform_to_ros(x) for x in bus.transforms]
    return out


class FoxgloveMsgsFrameTransformsMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from foxglove_msgs.msg import FrameTransforms as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return frame_transforms_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return frame_transforms_to_ros(bus)
